### PR TITLE
feat(s2i): Convert dockerfile to s2i image

### PR DESCRIPTION
**Pull request by mimotej**
Adds [S2I](https://docs.openshift.com/container-platform/4.10/openshift_images/using_images/using-s21-images.html) to the service catalog instead of `Dockerfile`. Since [backstage needs](https://backstage.io/docs/getting-started/running-backstage-locally#prerequisites) `yarn` to work (when `npm` is used there is bunch of errors related to `yarn` functionality) I had to create `.s2i/bin/assemble` script which installs `yarn` and installs dependencies using this package manager. I have also edited `start` script to run both FE and BE when image is started.

### CODE REVIEW

**Review by SamoKopecky — COMMENTED** [review]
Looks really good! Just have one small suggestion.

File: packages/backend/build-image.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# Creates temporary folder for the build
+tmp_dir=$(mktemp -d -t service-catalog-XXXXXXXXXX)
+
+
+echo "---> Run s2i build"
+s2i build . ${NODEJS_BASE_IMAGE} --as-dockerfile ${tmp_dir}/Containerfile
+cd $tmp_dir
+
+echo "---> Create image"
+podman build -t source-catalog .
```
[the remark this answers is not in the document]
Comment (mimotej) [reply packages/backend/build-image.sh]: Missed that :man_facepalming:. Thanks fixed :+1:

File: .s2i/bin/assemble [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+npm install -g yarn
+
+set -e
+safeLogging () {
+    if [[ $1 =~ http[s]?://.*@.*$ ]]; then
+        echo $1 | sed 's/^.*@/redacted@/'
+    else
+        echo $1
+    fi
+}
```
Comment (tumido) [inline .s2i/bin/assemble:12]: Dead code, not used

File: README.md [hunk 3]
```diff
@@ -93,9 +93,12 @@ cd manifests/overlays/dev && kustomize edit set image quay.io/operate-first/serv
...
-The image can be built using:
+### S2I image
+This repository uses s2i to create images. To create image run script located in `packages/backend`:
 ```sh
-yarn build-image
+./packages/backend/build-image.sh
+```
+This will generate an image called `service-catalog`, which then can be run with:
+```sh
+podman run -it --network host --env "backend_secret=$(node -p 'require("crypto").randomBytes(24).toString("base64")')" service-catalog:latest
```
Comment (tumido) [inline README.md:103]:
```suggestion
podman run -it -p 7007:7007 --env "backend_secret=$(node -p 'require("crypto").randomBytes(24).toString("base64")')" service-catalog:latest
```

Port 7007 works just fine for me...

File: package.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
   },
   "scripts": {
     "dev": "concurrently \"yarn start\" \"yarn start-backend\"",
-    "start": "yarn workspace app start",
+    "start": "yarn workspace app start & yarn workspace backend start",
```
Comment (tumido) [inline package.json:10]: I don't understand this change. Doesn't it start a dev server? Why don't we use the same command/entrypoint as is in the `Dockerfile`?

File: .s2i/bin/assemble [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+fix-permissions ./
+
+
+echo "---> Installing all dependencies"
+NODE_ENV=development yarn install
+
+#do not fail when there is no build script
+echo "---> Building in production mode"
+yarn run build
+
```
Comment (tumido) [inline .s2i/bin/assemble:33]: You should clean up all your dev dependencies after build so the image is as small as possible.

**Review by tumido — CHANGES_REQUESTED** [review]
We're on a good path here, some changes needed. This is just a quick review what I've noticed on the first glance.

File: packages/backend/Dockerfile [hunk 6]
```diff
@@ -1,43 +0,0 @@
...
-
-WORKDIR /app
-COPY --from=packages /app .
-
-RUN yarn install --frozen-lockfile --network-timeout 600000 && rm -rf "$(yarn cache dir)"
-
-COPY . .
-
-RUN yarn tsc
-RUN yarn --cwd packages/backend build
```
Comment (tumido) [inline packages/backend/Dockerfile:23]: This is missing from `assemble`, backend is not built at all

**Review by tumido — CHANGES_REQUESTED** [review]

File: .s2i/bin/assemble (hunk 2 above)
Comment (mimotej) [reply .s2i/bin/assemble]: Resolved in latest commit.

File: .s2i/bin/assemble (hunk 5 above)
Comment (mimotej) [reply .s2i/bin/assemble:22]: Changed it to only install production dependencies

File: README.md (hunk 3 above)
Comment (mimotej) [reply README.md]: After reinstall of podman it started working no idea what was wrong :man_shrugging:. Resolved

File: package.json (hunk 4 above)
Comment (mimotej) [reply package.json]: Yes, this wasn't good solution. Added `.s2i/bin/run` where I am using same entrypoint as is in the `Dockerfile`

File: .aicoe-ci.yaml [hunk 7]
```diff
@@ -1,2 +1,8 @@
 ---
 check: []
+build:
+  build-strategy: Source
+  base-image: registry.access.redhat.com/ubi9/nodejs-16:latest
```
Comment (tumido) [inline .aicoe-ci.yaml:5]: Have you tried `nodejs-16-minimal`?

File: .dockerignore [hunk 8]
```diff
@@ -1,9 +0,0 @@
-.git
```
Comment (tumido) [inline .dockerignore:1]: This should not be removed, it should be rather renamed to `.s2iignore` and used to shrink the image footprint.

File: .dockerignore [hunk 9]
```diff
@@ -1,9 +0,0 @@
...
-node_modules
-packages/*/node_modules
-plugins
-manifests
-dist-types
-.aicoe-ci.yaml
-.pre-commit-config.yaml
-.prow.yaml
```
Comment (tumido) [inline .dockerignore:9]: add `.gitignore`

**Review by tumido — CHANGES_REQUESTED** [review]
Very good progress, however I think the image size can be adjusted even a tiny bit more.

File: .dockerignore (hunk 8 above)
Comment (tumido) [reply .dockerignore:1]: ^ meaning the whole `.dockerignore` file, not the `.git` line only.

Comment (tumido) [reply .dockerignore:1]: See https://github.com/openshift/source-to-image#build-workflow 

image

Comment (mimotej) [reply .dockerignore]: Added in the latest commit :+1:

File: .dockerignore (hunk 9 above)
Comment (mimotej) [reply .dockerignore:9]: Resolved in latest commit

**Review by tumido — APPROVED** [review]
/lgtm
/approve

nice work :+1: :100:

File: .aicoe-ci.yaml (hunk 7 above)
Comment (mimotej) [reply .aicoe-ci.yaml]: This indeed decreseases size of the image size without using minimal image is 1.01 GB and with minimal it is only 583 MB. But using minimal image leads to error shown on the picture below.
image

tldr. From what I have gather it is an error made by ssh2 package when it is trying to optimize cipher ordering based on HW. More about it [here](https://github.com/backstage/backstage/issues/7523#issuecomment-945089084). It is used by dev dependency which is in the `yarn install --prodcution` removed so it should be safe to ignore
